src/codegen/compile.ml

```diff
@@ -9951,214 +9847,185 @@ let compile_binop env t op : SR.t * SR.t * G.t =
-            G.i (Binary (Wasm.Values.I32 I32Op.And)) ^^
-            G.if1 I32Type
+            get_n ^^ compile_unboxed_one ^^ compile_comparison I64Op.LeS ^^
+            get_n ^^ compile_unboxed_const (-1L) ^^ compile_comparison I64Op.GeS ^^
+            G.i (Binary (Wasm_exts.Values.I64 I64Op.And)) ^^
+            E.if1 I64Type
               begin
-                get_n ^^ compile_unboxed_zero ^^ G.i (Compare (Wasm.Values.I32 I32Op.LtS)) ^^
-                G.if1 I32Type
+                get_n ^^ compile_unboxed_zero ^^ compile_comparison I64Op.LtS ^^
+                E.if1 I64Type
                   begin
                     (* -1 ** (1+exp) == if even (1+exp) then 1 else -1 *)
-                    get_exp ^^ compile_unboxed_one ^^ G.i (Binary (Wasm.Values.I32 I32Op.And)) ^^
-                    G.if1 I32Type
-                      get_n
-                      compile_unboxed_one
+                    get_exp ^^ compile_unboxed_one ^^ G.i (Binary (Wasm_exts.Values.I64 I64Op.And)) ^^
+                    E.if1 I64Type
+                      (get_n ^^ TaggedSmallWord.msb_adjust ty)
+                      (compile_unboxed_const
+                        Int64.(shift_left one (to_int (TaggedSmallWord.shift_of_type ty))))
                   end
-                  get_n (* n@{0,1} ** (1+exp) == n *)
+                  (get_n ^^ TaggedSmallWord.msb_adjust ty) (* n@{0,1} ** (1+exp) == n *)
               end
               begin
-                get_exp ^^ compile_unboxed_const 32l ^^
-                G.i (Compare (Wasm.Values.I32 I32Op.GeU)) ^^ then_arithmetic_overflow env ^^
-                signed_dynamics get_n ^^ compile_sub_const 31l ^^
-                get_exp ^^ TaggedSmallWord.lsb_adjust Type.Int32 ^^ G.i (Binary (Wasm.Values.I32 I32Op.Mul)) ^^
-                compile_unboxed_const (-62l) ^^
-                G.i (Compare (Wasm.Values.I32 I32Op.LtS)) ^^ then_arithmetic_overflow env ^^
-                get_n ^^ G.i (Convert (Wasm.Values.I64 I64Op.ExtendSI32)) ^^
-                get_exp ^^ G.i (Convert (Wasm.Values.I64 I64Op.ExtendSI32)) ^^
-                Word64.compile_unsigned_pow env ^^
-                set_res ^^ get_res ^^ get_res ^^ enforce_32_signed_bits env ^^
-                get_res ^^ G.i (Convert (Wasm.Values.I32 I32Op.WrapI64))
+                let overflow_type = match ty with
+                | Type.Int32 -> Type.Int64 
+                | Type.(Int8 | Int16) -> Type.Int32 
+                | _ -> assert false in
+                let overflow_type_bits = TaggedSmallWord.bits_of_type overflow_type in
+                let overflow_boundary = -Int.(sub (mul overflow_type_bits 2) 2) in
+                get_exp ^^ compile_unboxed_const 64L ^^
+                compile_comparison I64Op.GeU ^^ then_arithmetic_overflow env ^^
+                signed_dynamics get_n ^^ compile_sub_const (Int64.of_int (Int.sub bits 1)) ^^
+                get_exp ^^ TaggedSmallWord.msb_adjust ty ^^ TaggedSmallWord.lsb_adjust Type.Int32 ^^
```

Comment:
This msb_adjust followed bu lsb_adjust with a different type description (ty vs Type.Int32) smells wrong to me. Can you explain it?

Reply:
Thanks a lot for noticing. This indeed looks wrong. I believe it should be simply `get_exp`.
The `TaggedSmallWord.lsb_adjust ty` is already applied at the beginning of the function before `set_exp` is called.

Very strange that no test case fails (considering that we quite a bunch of test cases for the "power of" function). It would be a too relaxed boundary check if I am not mistaken. I tried to reproduce but was not successful.

Reply:
I fixed it.